wpilib/wpilib/notifier.py

```diff
@@ -0,0 +1,124 @@
+# validated: 2017-12-24 DS 8b7aa61091df edu/wpi/first/wpilibj/Notifier.java
+#----------------------------------------------------------------------------
+# Copyright (c) FIRST 2016-2017. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+#----------------------------------------------------------------------------
+
+import threading
+
+import hal
+
+from .resource import Resource
+from .robotcontroller import RobotController
+
+class Notifier:
+    
+    def __init__(self, run: callable):
+        
+        #: The lock for the process information.
+        self._processLock = threading.RLock()
+        
+        # Notifier handle
+        self._notifier = hal.initializeNotifier()
+        
+        #: The time, in microseconds, at which the corresponding handler should be
+        #: called. Has the same zero as Timer.getFPGATime().
+        self._expirationTime = 0
+        
+        #: The handler passed in by the user which should be called at the
+        #: appropriate interval.
+        self._handler = run
+        
+        # Whether we are calling the handler just once or periodically.
+        self._periodic = False
+        
+        #: If periodic, the period of the calling; if just once, stores how long it
+        #: is until we call the handler.
+        self._period = 0
+        
+        #: The thread waiting on the HAL alarm
+        self._thread = threading.Thread(target=self._thread, daemon=True)
+        self._thread.start()
+        
+        # python-specific
+        Resource._add_global_resource(self)
+    
+    # python-specific: this is called finalize in Java
+    def free(self) -> None:
+        handle, self._notifier = self._notifier, None
```

Comment:
not that I can tell that whether it matters, but this is not an atomic operation